GliaWidgets/Sources/CallVisualizer/CallVisualizer.swift

```diff
@@ -131,31 +134,41 @@ extension CallVisualizer {
 
 extension CallVisualizer {
     func startObservingInteractorEvents() {
-        environment.interactorProviding()?.addObserver(self) { [weak self] event in
+        interactorSubscription = environment.interactorPublisher
+            .sink { [weak self] newInteractor in
+                guard let self else { return }
+                self.handleInteractorEvents(newInteractor)
+            }
+    }
+
+    func handleInteractorEvents(_ newInteractor: Interactor?) {
+        self.activeInteractor?.removeObserver(self)
+        self.activeInteractor = newInteractor
+        newInteractor?.addObserver(self) { [weak self] event in
+            guard let self else { return }
             if case .stateChanged(.ended(.byOperator)) = event,
-               let endedEngagement = self?.environment.interactorProviding()?.endedEngagement,
+               let endedEngagement = activeInteractor?.endedEngagement,
                endedEngagement.source == .callVisualizer {
-                self?.endSession()
-                self?.environment.log.prefixed(Self.self).info("Call visualizer engagement ended")
+                self.endSession()
```

Comment:
I think removing `self` here will ensure that compiler will let us know about how we capture `self`.
```suggestion
                self.endSession()
```